File: app/src/main/java/com/doapps/habits/helper/SimpleItemTouchHelperCallback.java
package com.doapps.habits.helper;

import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

import com.doapps.habits.models.MovableList;

public class SimpleItemTouchHelperCallback extends ItemTouchHelper.Callback {

    private final MovableList mAdapter;

    public SimpleItemTouchHelperCallback(final MovableList adapter) {
        mAdapter = adapter;
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return true;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return true;
    }

    @Override
    public int getMovementFlags(final RecyclerView recyclerView, final RecyclerView.ViewHolder viewHolder) {
        final int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
        final int swipeFlags = ItemTouchHelper.START | ItemTouchHelper.END;
        return ItemTouchHelper.Callback.makeMovementFlags(dragFlags, swipeFlags);
    }

    @Override
    public boolean onMove(final RecyclerView recyclerView, final RecyclerView.ViewHolder viewHolder,
                          final RecyclerView.ViewHolder target) {
        mAdapter.onItemMove(viewHolder.getAdapterPosition(), target.getAdapterPosition());
        return true;
    }

    @Override
    public void onSwiped(final RecyclerView.ViewHolder viewHolder, final int direction) {
        mAdapter.onItemDismiss(viewHolder.getAdapterPosition());
    }

}

File: app/src/main/java/com/doapps/habits/helper/SimpleItemTouchHelperCallback.kt
package com.doapps.habits.helper

import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper

import com.doapps.habits.adapter.IMovableListAdapter

class SimpleItemTouchHelperCallback(private val mAdapter: IMovableListAdapter) : ItemTouchHelper.Callback() {
  override fun isLongPressDragEnabled() = true
  override fun isItemViewSwipeEnabled() = true

  /**
   * Defines movement directions
   */
  override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
    val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
    val swipeFlags = ItemTouchHelper.START or ItemTouchHelper.END
    return ItemTouchHelper.Callback.makeMovementFlags(dragFlags, swipeFlags)
  }

  /**
   * Handles item move event
   */
  override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder,
                      target: RecyclerView.ViewHolder): Boolean {
    mAdapter.onItemMove(viewHolder.adapterPosition, target.adapterPosition)
    return true
  }

  /**
   * Handles item swipe event
   */
  override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) =
      mAdapter.onItemDismiss(viewHolder.adapterPosition)

}